libraries/stdlib/osl/stdosl.h

```diff
@@ -573,6 +573,226 @@ closure color volume_matte() BUILTIN;
+// This closure may be vertically layered over a base BSDF, where energy that is not reflected
+// will be transmitted to the base closure.
+//
+//  \param  N           Normal vector of the surface point beeing shaded.
+//  \param  albedo      Surface albedo.
+//  \param  roughness   Surface roughness [0,1].
+//
+closure color sheen_bsdf(normal N, color albedo, float roughness) BUILTIN;
+​
+// Adds an iridescent thin film layer over a microfacet base BSDF. This must be layered over
+// another base BSDF using the layer() closure, as this is a modifier and cannot be used as a
+// standalone closure.
+//
+// TODO:
+// - This might be better to represent as optional extra arguments on the closures that do support
+//   thin-film iridescence (dielectric_bsdf, conductor_bsdf and generalized_schlick_bsdf)?
+//
+//  \param  thickness   Thickness of the thin film.
+//  \param  ior         Refraction index of the thin film.
+//
+closure color thin_film_bsdf(float thickness, float ior) BUILTIN;
+​
+​
+// -------------------------------------------------------------//
+// EDF closures                                                 //
+// -------------------------------------------------------------//
+​
+// Constructs an EDF emitting light uniformly in all directions.
+//
+//  \param  emittance   Radiant emittance of light leaving the surface.
+//
+closure color uniform_edf(color emittance) BUILTIN;
+
+// Constructs an EDF emitting light inside a cone around the normal direction.
+//
+//  \param  emittance   Radiant emittance of light leaving the surface.
+//  \param  N           Cone direction vector.
+//  \param  inner_angle Angle of inner cone where emission falloff starts.
+//  \param  outer_angle Angle of outer cone where emission goes to zero. If set to a smaller value 
+//                      than inner_angle no falloff will occur within the cone.
+//
+closure color conical_edf(color emittance, normal N, float inner_angle, float outer_angle) BUILTIN;
+​
+​
+// -------------------------------------------------------------//
+// VDF closures                                                 //
+// -------------------------------------------------------------//
+​
+// Constructs a VDF scattering light for a participating medium, based on the Henyey-Greenstein
+// phase function. Forward, backward and uniform scattering is supported and controlled by the
+// anisotropy input.
+//
+//  \param  albedo      Volume single scattering albedo.
+//  \param  extinction  Volume extinction coefficient.
+//  \param  anisotropy  Scattering anisotropy [-1,1]. Negative values give backwards scattering, positive values give forward scattering, 
+//                      and 0.0 gives uniform scattering.
+//  \param  ior         Optional float parameter for refraction index of a homogeneous medium.
+//  \param  priority    Optional int parameter for priority of a homogeneous medium (for nested dielectrics).
```

Comment:
Is it too verbose if we make these formal parameters instead of optional? But I admit we've been using optional args for new features added after an API has already been published, so when I see a new API I tend to visualize everything as formal param. Maybe because OSL doesn't have syntax for declaring keyword args and you have to read the comments.

Reply:
I agree it's not ideal one has to read the docs to know what the full parameterization is. On the other hand the ior and priority parameters would only be used in the case of a dielectric / homogenous medium, so it's a bit awkward having to set values for those in other contexts.

Does OSL support default values on parameters of built-in functions? If so we could make them formal parameters with default values, so you only set them when needed.

If not I suppose another option is to define a wrapper function for homogenous media with the full parameterization declared:

```
closure color media_vdf(color albedo, color extinction, float anisotropy, float ior, int priority)
{
    return anisotropic_vdf(albedo, extinction, anisotropy, "ior", ior, "priority", priority);
}
```

Otherwise, as you mentioned, the most elegant solution would be to add a keyword args syntax for OSL. I think I read something about this topic in the meeting notes from last weeks OSL meeting.

Getting off topic here but just wanted to mention that in Nvidia's MDL there is such a syntax (they call it _named arguments_ ) that looks like this:

```
float foo(float a = 7.0, float b = 42.0)
{
    return a + b;
}

float x1 = foo()             // equals 49.0
float x2 = foo(4.0, 5.0)     // equals  9.0
float x3 = foo(a:4.0, b:5.0) // equals  9.0
float x4 = foo(b:6.0, a:4.0) // equals 10.0
float x5 = foo(a:3.0)        // equals 45.0
float x6 = foo(b:8)          // result 15.0
```

Reply:
I like that syntax a lot. We talked about it at the last OSL TSC meeting, and I'm in favor of adding it.

Reply:
Meanwhile, your wrapper proposal sounds good to me.

That syntax is something we should have done a long time ago. The current ```"key", value);``` thing is ugly. But I would choose ```key = value);``` instead of ```key:value);``` if you give me the option :)

Reply:
If we are adding the media_vdf wrapper (or even a separate built-in closure for this) would it then make sense to use the same parameterization as for subsurface_bssrdf, with the artist-friendly transmission_depth and transmission_color? For a closure that is dedicated for dielectric homogenous media, this seems much easier to control.